.buildkite/autogenerate_pipeline.py

```diff
@@ -0,0 +1,230 @@
+
+        env_var = None
+        platform = self.agents.get('platform')
+
+        if platform:
+            if platform == 'x86_64.metal' and X86_AGENT_TAGS:
+                env_var = X86_AGENT_TAGS
+            if platform == 'arm.metal' and AARCH64_AGENT_TAGS:
+                env_var = AARCH64_AGENT_TAGS
+
+        target = self.agents
+        self._env_change_config(test_name, env_var, target, override=True)
+
+    def _env_add_docker_config(self, test_name):
+        """
+        Specify additional configuration for the docker plugin using the
+        `DOCKER_PLUGIN_CONFIG` environment variable.
+        """
+
+        target = self.plugins[0][f"docker#{DOCKER_PLUGIN_VERSION}"]
+        self._env_change_config(test_name, DOCKER_PLUGIN_CONFIG, target)
+
+    def build(self, input):
+        """
+        Build a Buildkite step using the `input` configuration that must
+        specify some mandatory keys and can also provide optional ones.
+        Further configuration from environment variables may be added.
+        """
+
+        test_name = input.get('test_name')
+        command = input.get('command')
+        platform = input.get('platform')
+        docker = input.get('docker_plugin')
+
+        # Mandatory keys.
+        assert test_name, "Step is missing test name."
+        platform_string = f"-{platform}" if platform else ""
+        self.label = f"{test_name}{platform_string}"
+
+        assert command, "Step is missing command."
+        if "{target_platform}" in command:
+            assert platform,\
+                "Command requires platform, but platform is missing."
+            command = command.replace(
+                "{target_platform}", platform
+            )
+        self.command = command
+
+        # Optional keys.
+        self._set_platform(platform)
+        self._add_docker_config(docker)
+
+        # Override/add configuration from environment variables.
+        self._env_override_agent_tags(test_name)
+        self._env_add_docker_config(test_name)
+
+        return self.__dict__
+
+
+class BuildkiteConfig:
+    """
+    This builds the final Buildkite configuration from the json input
+    using BuidkiteStep objects. The output is a dictionary that can
+    be put into yaml format by the pyyaml package.
+    """
+
+    def build(self, input):
+        """ Build the final Buildkite configuration fron the json input. """
+
+        self.steps = []
+        tests = input.get('tests')
+        assert tests, "Input is missing list of tests."
+
+        for test in tests:
+            platforms = test.get('platform')
+
+            if not platforms:
+                platforms = [None]
+
+            for platform in platforms:
+                step_input = copy.deepcopy(test)
+                step_input['platform'] = platform
+
+                step = BuildkiteStep()
+                step_output = step.build(step_input)
+                self.steps.append(step_output)
+
+        return self.__dict__
```

Comment:
Can we leave a comment explaining what __dict__ does?

Reply:
Done. Also replaced `self.__dict__` with `vars(self)`. This seems to be the more "Pythonic" way because it avoids calling magic methods such as `__dict__`.